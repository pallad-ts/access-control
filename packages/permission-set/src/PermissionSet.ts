import {EntityRef, Factory} from "@pallad/entity-ref";
import {AccessQuery, Policy} from "@pallad/access-control";
import {SubjectDefinition} from "./types";

export class PermissionSet<T extends [string, unknown]> {
    private entriesActionToSubjectPredicates = new Map<string, Set<SubjectDefinition.TypeGuard<any>>>();

    readonly type!: T;

    constructor(readonly subjectType: string) {

    }

    canHandleQuery({action, subject}: AccessQuery) {
        return this.isValidSubject(subject) && this.isValidSubjectForAction(action, subject.wrappedSubject);
    }

    private isValidSubjectForAction(action: string, subject: unknown) {
        const entityRefFactoriesForPrivilege = this.entriesActionToSubjectPredicates.get(action);
        return Array.from(entityRefFactoriesForPrivilege ?? []).some(typeGuard => typeGuard(subject));
    }

    private isValidSubject(subject: unknown): subject is PermissionSet.Subject<T[1]> {
        // eslint-disable-next-line no-null/no-null
        return typeof subject === 'object' && subject !== null && 'subjectType' in subject && (subject as any).subjectType === this.subjectType;
    }

    wrapSubjectWithSubjectType<T>(wrappedSubject: T): PermissionSet.Subject<T> {
        return Object.freeze({
            wrappedSubject,
            subjectType: this.subjectType
        });
    }

    /**
     * Adds combination of action and subject definition to set of combinations that current permission set can handle
     */
    addEntry<
        TAction extends string,
        TSubjectDefinition extends SubjectDefinition
    >(action: TAction, subjectDefinition: TSubjectDefinition) {
        if (!this.entriesActionToSubjectPredicates.has(action)) {
            this.entriesActionToSubjectPredicates.set(action, new Set());
        }
        this.entriesActionToSubjectPredicates.get(action)!.add(createSubjectPredicateFromSubjectDefinition(subjectDefinition));
        return this;
    }

    /**
     * Returns a policy for provided policy function.
     * Provided policy function is going to ba called only for access query that is matched agains current permission set.
     */
    createPolicy<TPrincipal>(
        policyFunction: Policy<PermissionSet.PolicyAccessQueryInput<TPrincipal, T>>
    ): Policy<PermissionSet.PolicyAccessQueryOutput<TPrincipal, T>> {
        return (query: PermissionSet.PolicyAccessQueryOutput<TPrincipal, T>) => {
            if (this.canHandleQuery(query)) {
                return policyFunction({
                    action: query.action,
                    subject: query.subject.wrappedSubject,
                    principal: query.principal
                } as any);
            }
        }
    }
}


function createSubjectPredicateFromSubjectDefinition(subjectDefinition: SubjectDefinition): SubjectDefinition.TypeGuard<any> {
    if (isEntityRefFactory(subjectDefinition)) {
        return subjectDefinition.is;
    }

    if (isClassFunction(subjectDefinition)) {
        return (value: unknown): value is any => value instanceof subjectDefinition;
    }

    if (typeof subjectDefinition === 'function') {
        return subjectDefinition;
    }

    throw new Error('Invalid subject definition provided. Only EntityRefFactory, class or predicate function allowed');
}

function isEntityRefFactory(subjectDefinition: SubjectDefinition): subjectDefinition is Factory<string, any> {
    return typeof subjectDefinition === 'function' && 'typeName' in subjectDefinition && 'is' in subjectDefinition;
}

function isClassFunction(func: unknown): func is SubjectDefinition.Constructor<any> {
    return typeof func === 'function' && Function.prototype.toString.call(func).startsWith('class ');
}

export namespace PermissionSet {
    export type PolicyAccessQueryInput<TPrincipal, T extends [string, unknown]> =
        T extends [infer TAction extends string, infer TSubject] ?
            AccessQuery<TPrincipal, TAction, TSubject>
            : AccessQuery<TPrincipal>;

    export type PolicyAccessQueryOutput<TPrincipal, T extends [string, unknown]> =
        T extends [infer TAction extends string, infer TSubject] ?
            AccessQuery<TPrincipal, TAction, Subject<TSubject>>
            : AccessQuery<TPrincipal>;

    export interface Subject<TRealSubject> {
        readonly subjectType: string;
        readonly wrappedSubject: TRealSubject;
    }

    export interface AccessQueryParams<TAction extends string, TEntityRef extends EntityRef<string, {}>> {
        action: TAction;
        subject: Subject<TEntityRef>;
    }

    export type SubjectPredicate = (value: unknown) => boolean;
}




