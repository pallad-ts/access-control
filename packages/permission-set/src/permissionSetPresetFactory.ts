import {PermissionSet} from "./PermissionSet";
import {InferSubjectFromDefinition, SubjectDefinition} from "./types";

export function permissionSetPresetFactory<T extends Record<string, readonly [string, SubjectDefinition]>>(name: string, methods: T) {

    const permissionSet = new PermissionSet(name);
    const result: any = {
        permissionSet,
    };

    for (const [method, [action, subjectDefinition]] of Object.entries(methods)) {
        permissionSet.addEntry(action, subjectDefinition);
        result[method] = (subject: any): PermissionSet.AccessQueryParams<string, any> => {
            return {action, subject: permissionSet.wrapSubjectWithSubjectType(subject)};
        }
    }

    return result as PermissionSetPreset<T>;
}


export type PermissionSetPreset<T extends Record<string, readonly [string, SubjectDefinition]>> = {
    permissionSet: PermissionSet<MethodsToAccessQueryParamsUnion<T[keyof T]>>,
} & {
    [TKey in keyof T]: (ref: InferSubjectFromDefinition<T[TKey][1]>) => PermissionSet.AccessQueryParams<T[TKey][0], InferSubjectFromDefinition<T[TKey][1]>>
}

export type MethodsToAccessQueryParamsUnion<T extends readonly [string, SubjectDefinition]> =
    T extends readonly [infer TAction extends string, infer TSubjectDefinition extends SubjectDefinition] ?
        [TAction, InferSubjectFromDefinition<TSubjectDefinition>]
        : never;
