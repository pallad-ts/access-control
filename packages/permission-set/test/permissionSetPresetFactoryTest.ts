import {createFactory} from "@pallad/entity-ref";
import {permissionSetPresetFactory} from "@src/permissionSetPresetFactory";
import {IsExact, assert} from "conditional-type-checks";
import {PermissionSet} from "@src/PermissionSet";
import {ORGANIZATION_REF, organizationFactory, PRODUCT_REF, productFactory} from "./fixtures";
import {Subject} from "@pallad/access-control";

describe('permissionSetPresetFactory', () => {
    const FOO_BAR = {foo: 'bar'} as const;
    const factory = permissionSetPresetFactory('product', {
        canUpdate: ['update', productFactory] as const,
        canDelete: ['delete', productFactory] as const,
        canCreate: ['create', organizationFactory] as const,
        canTruncate: ['truncate', Subject.Global] as const,
        canActivate: ['activate', (value: unknown): value is typeof FOO_BAR => {
            return value === FOO_BAR;
        }] as const
    });

    it('provided methods are assigned to final object and those methods return access control query params with action and subject', () => {
        expect(factory.canDelete(PRODUCT_REF))
            .toMatchSnapshot();

        expect(factory.canCreate(ORGANIZATION_REF))
            .toMatchSnapshot();

        expect(factory.canUpdate(PRODUCT_REF))
            .toMatchSnapshot();
    });

    it('types', () => {
        type ProductEntityRef = ReturnType<typeof productFactory>;
        type OrganizationEntityRef = ReturnType<typeof organizationFactory>;

        type Factory = typeof factory;

        assert<IsExact<Parameters<Factory['canCreate']>[0], OrganizationEntityRef>>(true);
        assert<IsExact<Parameters<Factory['canDelete']>[0], ProductEntityRef>>(true);
        assert<IsExact<Parameters<Factory['canUpdate']>[0], ProductEntityRef>>(true);

        assert<IsExact<Factory['permissionSet'], PermissionSet<
            ['update', ProductEntityRef] | ['delete', ProductEntityRef] | ['create', OrganizationEntityRef] | ['truncate', Subject.Global] | ['activate', typeof FOO_BAR]
        >>>(true);
    });

    it('returned permission set is preconfigured with all methods', () => {
        expect(factory.permissionSet.canHandleQuery({
            ...factory.canUpdate(PRODUCT_REF),
            principal: {}
        })).toBe(true);

        expect(factory.permissionSet.canHandleQuery({
            ...factory.canDelete(PRODUCT_REF),
            principal: {}
        })).toBe(true);

        expect(factory.permissionSet.canHandleQuery({
            ...factory.canCreate(ORGANIZATION_REF),
            principal: {}
        })).toBe(true);

        expect(factory.permissionSet.canHandleQuery({
            ...factory.canTruncate(Subject.Global.INSTANCE),
            principal: {}
        })).toBe(true);

        expect(factory.permissionSet.canHandleQuery({
            ...factory.canActivate(FOO_BAR),
            principal: {}
        })).toBe(true);
    });
});
