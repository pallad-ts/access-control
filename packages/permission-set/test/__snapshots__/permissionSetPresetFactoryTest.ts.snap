// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`permissionSetPresetFactory provided methods are assigned to final object and those methods return access control query params with action and subject 1`] = `
{
  "action": "delete",
  "subject": {
    "subjectType": "product",
    "wrappedSubject": EntityRef {
      "data": {
        "id": "productId",
        "organizationId": "organizationId",
      },
      "type": "product",
    },
  },
}
`;

exports[`permissionSetPresetFactory provided methods are assigned to final object and those methods return access control query params with action and subject 2`] = `
{
  "action": "create",
  "subject": {
    "subjectType": "product",
    "wrappedSubject": EntityRef {
      "data": {
        "id": "organizationId",
      },
      "type": "organization",
    },
  },
}
`;

exports[`permissionSetPresetFactory provided methods are assigned to final object and those methods return access control query params with action and subject 3`] = `
{
  "action": "update",
  "subject": {
    "subjectType": "product",
    "wrappedSubject": EntityRef {
      "data": {
        "id": "productId",
        "organizationId": "organizationId",
      },
      "type": "product",
    },
  },
}
`;
